Added emerge status module

AnwariasEu opened the pull request:
Added a module that displays current emerge information.
Multiple format options are available (documented in the module file).

User needs to have read access to the log file in /var/log/emerge.log. This can be acomplished by the Gentoo 'portage' group.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,144 @@
+                for line in fp:
+                    input_data.append(line)
+        except IOError as err:
+            ret['err'] = True
```
This could be better off with `self.py3.error()` instead. I don't know how often this can happen so if it's not something very often, then we could skip `Try/Except` instead. The exception would be shown on the bar.

We also could check for binary `emerge` in `post_config_hook` so this can't happen on other distributions.

AnwariasEu replied:
This can happen if the file does not exist e.g. the module is used on system where there is no portage and therefore no emerge.log.
But it can also happen the user has no read access to the file, which happens in case the user is not part of the 'portage' group.

Error handling can definitely be improved.

lasers replied:
We could make a `STRING_NOT_INSTALLED` (see `moc.py`) so this would only run on machines with installed `emerge` == portage (?). I wonder if people who uses `emerge` regularly may already add themselves to `portage` group? If not, we could make a small note about `portage` group. Then we might not have to worry about error handling anymore. And if there was one, users can read the exception on the bar instead of custom error string.

AnwariasEu replied:
emerge is the main binary of the portage package manager.
Checking for emerge binary implemented in 42e811f.
Error handling addressed in  e9d9ab8.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,144 @@
+        ret['err'] = False
+        ret['err_text'] = ""
+        ret['category'] = None
+        ret['pkg'] = None
```
See if we can simplify things enough to start with just `ret = {}`.

AnwariasEu replied:
I guess this change only makes sense if the undocumented modifiers like `\?if`  are used?

lasers replied:
This is to kill lines. The format can use `[{example}]` to make non-existent  placeholders and `None` placeholders go away. It looks like some values would be filled one way or other in the loop.

Or you could leave `ret = {}` and move the others down to `*** terminating line`. Use `''` instead of `None` (to make it look nice).

AnwariasEu commented:
First of all thanks for the feedback.
I'm pretty new to contributing to other work. 
I made some fixes in more commits, hopefully i will submit them correctly.

> * (Imo) I think it It would be nice to add prefix in the format too like DPMS, XSCR, UPD, etc, eg EMG {current}/{total} or whatever you think is better.

You mean in den default format?
I chose no prefix by personal preference, but there can definitely a prefix in the default.

> * Can we get some example emerge.log too?

Short except, want me to pastebin more?
```1484779948: Started emerge on: Jan 18, 2017 22:52:27
1484779948:  *** emerge --quiet --ask --tree --verbose sys-kernel/gentoo-sources
1484779959:  >>> emerge (1 of 2) sys-devel/bc-1.06.95-r1 to /
1484779962:  === (1 of 2) Cleaning (sys-devel/bc-1.06.95-r1::/usr/portage/sys-devel/bc/bc-1.06.95-r1.ebuild)
1484779962:  === (1 of 2) Compiling/Merging (sys-devel/bc-1.06.95-r1::/usr/portage/sys-devel/bc/bc-1.06.95-r1.ebuild)
1484779973:  === (1 of 2) Merging (sys-devel/bc-1.06.95-r1::/usr/portage/sys-devel/bc/bc-1.06.95-r1.ebuild)
1484779974:  >>> AUTOCLEAN: sys-devel/bc:0
1484779975:  === (1 of 2) Post-Build Cleaning (sys-devel/bc-1.06.95-r1::/usr/portage/sys-devel/bc/bc-1.06.95-r1.ebuild)
1484779975:  ::: completed emerge (1 of 2) sys-devel/bc-1.06.95-r1 to /
1484779975:  >>> emerge (2 of 2) sys-kernel/gentoo-sources-4.4.39 to /
1484781033:  *** terminating.
1484818385: Started emerge on: Jan 19, 2017 09:33:04
1484818385:  *** emerge --ask sys-kernel/gentoo-sources
1484818412:  >>> emerge (1 of 1) sys-kernel/gentoo-sources-4.4.39 to /
1484818412:  === (1 of 1) Cleaning (sys-kernel/gentoo-sources-4.4.39::/usr/portage/sys-kernel/gentoo-sources/gentoo-sources-4.4.39.ebuild)
1484818413:  *** Finished. Cleaning up...
1484818413:  *** exiting unsuccessfully with status '1'.
1484818413:  *** terminating.
```

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,146 @@
+        try:
+            with open(self.emerge_log_file, 'r') as fp:
+                for line in fp:
+                    input_data.append(line)
```
Here's an idea. This might be better to use `lines = open(self.log_file).readlines()` and maybe do `[-1]` too for last line only and fix up if/else to work with the last line. Based on the logs, we could make a `status` out of it too... like this: `Cleaning 1/2`, `Emerging 1/2`, `Compiling 1/2`, `Merging 1/2`, `Post-Build Cleaning 1/2`.... if users want that over static `EMG 1/2`... It's a bit more fun. :-)

AnwariasEu replied:
Readlines implemented in 5fc185a.
Status implemented as action in 1d78b05.

AnwariasEu commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,144 @@
+    """
+    """
+    cache_timeout = 1
+    emerge_log_file = '/var/log/emerge.log'
```
The location does not change, i mostly used this option for debugging.
Addressed in 8057c2b.

AnwariasEu commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,146 @@
+
+        for line in input_data:
+            terminated = re.search("\*\*\* terminating\.", line)
+            if terminated:
```
Addressed in 53afb9b.

AnwariasEu commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,144 @@
+                    self.format, self.ret)
+        else:
+            if self.hide_if_zero:
+                response['full_text'] = ""
```
I personally think this a rather confusing formatting style and more likely to discourage users from using a custom format. Even more since the function is not, yet, documented.

lasers replied:
I think @tobes may have a branch to deal with this stuff. What is more is that the formatters can do more than just hiding zeroes... IIRC it can do `\?if=max_length 1`, `\?if=min_length 1`, `\?not_zero`, `\?if=placeholder`, `{placeholder:d}` (decimal), `{placeholder:.0f}` (rounding a float), and can grow more later. See `./tests/test_formatter.py` for more possibilities. 

Maybe we will get it documented for `3.8` or `3.9`, but will be easier to fix this module in the `format` now rather than to try and fix it it later / deprecate.

Also, you can add `Examples: ` with `# hide if zero` comment too.

AnwariasEu commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,144 @@
+        tmp = os.popen("pgrep emerge").read()
+        pcount = tmp.count("\n")
+        if pcount > 0:
+            return True
```
Addressed in 590ed83. And fixed in 91d9e07.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,146 @@
+        response = {}
+        response['cached_until'] = self.py3.time_in(self.cache_timeout)
+        if self._emergeRunning():
+            self.ret = self._getProgress()
```
With `cache_timeout = 0`, we can just read the log instead.
* Skips running `i3-msg` every second.
* Skips running `pgrep` every second.

lasers replied:
Agreed. Check the log's last line instead of running `sh`+`i3-msg`+`pgrep` all times. More nice on your system resources.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,139 @@
+
+    def post_config_hook(self):
+        if not self.py3.check_commands('emerge'):
+            raise Exception("emerge not installed")
```
Look at `moc.py`. Make constant `STRING_NOT_INSTALLED` with string `not_installed`. The prefix name will be `emerge_status` which is a name of this module unless you want to change it to something else like... `emerge`. :-)

AnwariasEu replied:
Thanks for the explanation! Totally makes sense to keep things consistent.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,139 @@
+
+Format examples:
+     {current} / {total}'
+    Result: "3 / 14"
```
Wrap this part in `Examples: """ (docstrings) """`. See other modules for some examples. (ha)

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,139 @@
+        for line in input_data:
+            if "*** terminating." in line:
+                ret['current'] = "0"
+                ret['total'] = "0"
```
You can use real `0` and `None` everywhere instead of `"0"` and `"None"`.

AnwariasEu commented:
I have not been using a Gentoo setup lately.
I have fixed the string formatting (using the new modfiers) and will test it for a while/distribute it for testing.
Will probably update the branch in a week.

lasers commented:
_/me gives a friendly nudge_

AnwariasEu commented:
`Agreed. Check the log's last line instead of running sh+i3-msg+pgrep all times. More nice on your system resources.`  
Sadly this is not possible. We need to check by process id to see if emerge is running since the logfile is not deterministic if the user stops a running emerge process e.g. using ctrl+c. In that case the module would assume there is still a running emerge job (this behavior is even restart save).

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        response = {}
+        response['cached_until'] = self.py3.time_in(self.cache_timeout)
+        response['is_running'] = self._emergeRunning()
```
Putting unrelated things in `response` is misleading. Use `is_running = self._emergeRunning()` instead.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+        response = {}
+        response['cached_until'] = self.py3.time_in(self.cache_timeout)
+        response['is_running'] = self._emergeRunning()
+        ret = copy.deepcopy(self.ret_default)
```
Is this too much? Does `ret = self.ret_default` works fine?

AnwariasEu replied:
Since ret_default is a dictionary filled with references the assignment `ret = self.ret_default` will copy all references and not their actual values. As a result when setting e.g. ret['current'] we override the default value as well. I wanted to use ** here but this has only been introduced in python 3.5 and travis for 3.4 did not like it.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+            with open(emerge_log_file, 'r') as fp:
+                input_data = fp.readlines()
+        except IOError as err:
+            raise Exception(err)
```
If you want to show error, you can remove `try/except` and py3status will catch the error. Otherwise, I think this wouldn't work correctly without the log, so you probably want to remove `try/except`. It's either that or keep going with nothing.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+            'category': "",
+            'pkg': "",
+            'action': "",
+            'is_running': False}
```
I'm not sure if this makes sense, but at same time, it makes sense too. If emerge is not running, you're going to start passing in fake data. `total = 0` is not really `0` for next several days until we finally run `emerge`. Ideal behavior could be that we don't pass anything and the placeholders would display `{total}` instead... and we come up with something in the `format` to deal with that.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+
+        ret = {}
+
+        input_data = []
```
You might not need this as you already have one below.

AnwariasEu replied:
Maybe i can leave out definition of the `input_data` but `ret` needs to be be initialized somewhere as a dict. If i do so at beginning of the function or inside the if (right before the assignments of the values e.g. `ret['current']`) does not change that much.

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+            raise Exception("emerge not installed")
+
+    def _getProgress(self):
+        """
```
Python naming conventions for functions [[pep link]](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names), ie `_get_progress`. :+1:

lasers commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+        for line in input_data:
+            if "*** terminating." in line:
+                # Copy content of ret_default no only the references
+                ret = copy.deepcopt(self.ret_default)
```
Typo.

AnwariasEu commented:
> Here's what does not make sense.
>   hide_if_stopped does nothing.
>   prefix does nothing.
>    show_action does nothing.
>    show_pkg does nothing.

> If users want or don't want something, they can tweak everything in the format instead of limited customization boolean configs. We probably need to work on your format too.

I personally don't think format strings of that size are user friendly and easy to edit.
I want to offer the user a simple set of variables he can tweak to change the output of the module.

lasers commented:
Your current `format` is maybe the largest default `format` in py3status right now. It also looks unfriendly because of your set of variables. You're putting everything in the `format` then have users toggle on/off a set of variables instead of having users change things in the `format` (like many other modules).

We absolutely want users to change things in the `format`. Otherwise, we can keep getting pull requests that this user wanted to hide this... This user want to show this only if... This users want to start changing colors only if this is running... or to hide everything if a certain boolean is True.

We also want to offer users set of `format` examples instead where they can copy and paste what they like into their config... then if necessary, make small changes from there. Some of my examples can be found in `thunderbird_todos`, `coin_market`, `mail`, `keyboard_locks`, `loadavg`, `lm_sensors`, etc.

When it comes to customization, putting everything (if possible) in the `format` is easily more powerful than a set of static configs. Your set of static configs would become useless in an instant the users added the `format` to their config. For this to work, you require users to keep all those set of static configs in the `format` too... We have other thing for that instead... Placeholders.

If users can add colors, then they can add `{action}` too instead of toggling `show_action`.

(offtopic below) We already can do this with `percent = 1234.56789`.
* `{percent}` --> `1234.56789`
* `{percent:.3f}` --> `1234.567`
* `{percent:.2f}` --> `1234.57`
* `{percent:.0f}` --> `1235`
* `{percent:d}` --> `1234` <-- Maybe or `1235`.


There are other things we could add in the future too... such as allowing users to print both weather temperatures ie `format = '{weather_temp}°C and {weather_temp:c2f}°F'` assuming it's already in Celsius and `c2f` would convert Celsius to Fahrenheit. This is far more flexible than a boolean `Fahrenheit` config which would only can print one weather temperature in `format`.

AnwariasEu commented on py3status/modules/emerge_status.py:
```diff
@@ -0,0 +1,160 @@
+
+    def __init__(self):
+        self.ret_default = {
+            'current': 0,
```
Why put the initialization there?
Are there performance implications that favor using the `post_config_hook` instead of the builtin `__init__` ?

lasers replied:
For more information, see `Module init() method #399` PR.

https://github.com/ultrabug/py3status/blob/d395f2e57b95b1307eeed1c44d0bbc50825e7c3b/py3status/module.py#L130-L137

Tested just now... If we try to use `self.py3` helpers inside `__init__()`, we get this.
```bash
$ python static_string.py
Traceback (most recent call last):
  File "static_string.py", line 39, in <module>
    module_test(Py3status)
  File "py3status/py3status/module_test.py", line 67, in module_test
    module = module_class()
  File "static_string.py", line 22, in __init__
    self.py3.log('__init__: {}'.format(self.format))
AttributeError: 'Py3status' object has no attribute 'py3'
```
It's preferred too. All modules already had been updated to use `post_config_hook`.

AnwariasEu commented:
- Fixed the error message.
- Simplified the format string. I thinks its still not easy to read/write, but it matches the new formatting.
Perhaps it can be further simplified.
- Removed try/execpt when trying to open the file.
- Cache timeout is still missing.

ultrabug commented:
Thanks a lot @AnwariasEu 

I'm merging this now since you've been working on this for a while. I'm allowing myself to change some logic and QA on the module to be more gentle with disk I/O 

I've tested it, it's great!
